### PR TITLE
Masterchef deposit proof

### PR DESCRIPTION
**Pull request by gulshanvas**
This PR allows to confirm transfer of Validator shares using deposit method by providing receipt proof generate off-chain.

### CODE REVIEW

File: contracts/MasterChef.sol [hunk 1]
```diff
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.7.0;
+
+// import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+// import "@openzeppelin/contracts/math/SafeMath.sol";
+// import "./library/TransferHelper.sol";
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: Lets remove remove the commented imports.

File: contracts/MasterChef.sol [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        uint256 endBlock;
+        uint256 lastRewardBlock;  // Last block number that reward distribution occurs.
+        uint256 accRewardPerShare; // Accumulated Rewards per share, times 1e12
+        uint256 farmableSupply; // set in init, total amount of tokens farmable
+        uint256 numFarmers;
+    }
+    
+    /// @notice farm type id. Useful for back-end systems to know how to read the contract (ABI) 
+    /// as we plan to launch multiple farm types
+    uint256 public farmType = 1;
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: Not needed.

File: contracts/MasterChef.sol [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+   
+    address public farmGenerator;
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: Can be removed.

File: contracts/MasterChef.sol [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+    constructor( address _farmGenerator) public {
+        farmGenerator = _farmGenerator;
+    }
+
+    /**
+     * @notice initialize the farming contract. 
+     * This is called only once upon farm creation and the FarmGenerator ensures the farm has the correct paramaters
+     */
+    function init (
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: init method should be removed and initializations can be done in constructor.

File: contracts/MasterChef.sol [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        farmInfo.farmableSupply = _amount;
+    }
+
+    /**
+     * @notice Gets the reward multiplier over the given _from_block until _to block
+     * @param _from_block the start of the period to measure rewards for
+     * @param _to the end of the period to measure rewards for
+     * @return The weighted multiplier for the given period
+     */
+    function getMultiplier(uint256 _from_block, uint256 _to) public view returns (uint256) {
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: _from_block to _fromBlock ( camel case )

File: contracts/MasterChef.sol [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+            return;
+        }
+        uint256 multiplier = getMultiplier(farmInfo.lastRewardBlock, block.number);
+        uint256 tokenReward = multiplier.mul(farmInfo.blockReward);
+        farmInfo.accRewardPerShare = farmInfo.accRewardPerShare.add(tokenReward.mul(1e12).div(lpSupply));
+        farmInfo.lastRewardBlock = block.number < farmInfo.endBlock ? block.number : farmInfo.endBlock;
+    }
+
+    function deposit(
+        uint256 _pid,
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: Not needed ?

File: contracts/MasterChef.sol [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        );
+
+    }
+
+    
+    /**
+     * @notice deposit LP token function for msg.sender
+     * @param _amount the total deposit amount
+     */
+    function deposit_internal(address _user, uint256 _amount) internal {
```
Comment (ashwinYardi) [inline contracts/MasterChef.sol]: _depositInternal()

**Review by ashwinYardi — CHANGES_REQUESTED** [review]
Good one @gulshanvas ! Minor changes here and there. Also lets add deployment script for the ease of use.

File: contracts/MasterChef.sol (hunk 6 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Yes, not needed

File: contracts/MasterChef.sol (hunk 7 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Inspiration from recently writing code in JS :stuck_out_tongue: 
Thanks :smiley:

File: contracts/MasterChef.sol (hunk 5 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Done

File: contracts/MasterChef.sol (hunk 4 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Done

File: contracts/MasterChef.sol (hunk 3 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Done

File: contracts/MasterChef.sol (hunk 2 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Done

File: contracts/MasterChef.sol (hunk 1 above)
Comment (gulshanvas) [reply contracts/MasterChef.sol]: Done

**Review by ashwinYardi — APPROVED** [review]

**Pull request merged**